lib/routes/account.js

```diff
@@ -1021,50 +1048,78 @@ module.exports = function (
-        db.account(Buffer(uid, 'hex'))
-          .then(
-            function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', { email: account.email, uid: account.uid, locale: account.locale })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid)
+        if(code.length === 32) {
+          // Hex code length of 32 signifies a tokenId, perform login token verification
+          log.begin('Account.RecoveryLoginVerify', request);
+          db.verifyToken(code, {uid: uid})
+            .then(function () {
+            })
+            .done(
+              function () {
+                reply({})
+              },
+              reply
+            )
+        } else {
+          log.begin('Account.RecoveryEmailVerify', request)
+          db.account(Buffer(uid, 'hex'))
+            .then(
+              function (account) {
+                // If the account is already verified, they may be e.g.
+                // clicking a stale link.  Silently succeed.
+                if (account.emailVerified) {
+                  log.increment('account.already_verified')
+                  return true
+                }
+                if (!butil.buffersAreEqual(code, account.emailCode)) {
+                  throw error.invalidVerificationCode()
+                }
+                log.timing('account.verified', Date.now() - account.createdAt)
+                log.event('verified', {email: account.email, uid: account.uid, locale: account.locale})
+                log.increment('account.verified')
 
-              return db.verifyEmail(account)
-                .then(mailer.sendPostVerifyEmail.bind(
-                    mailer,
-                    account.email,
-                    {
-                      acceptLanguage: request.app.acceptLanguage
-                    }
-                  )
-                )
-            }
-          )
-          .done(
-            function () {
-              reply({})
-            },
-            reply
-          )
+                // send a push notification to all devices that the account changed
+                push.notifyUpdate(uid)
+
+                return db.sessions(account.uid)
+                  .then(function(sessions) {
+                    // TODO How to know which session is associated with the email verification?
+                    // Assuming last created session?
```

Comment:
I'm not sure this is a problem because I *think* we agreed that the account will be verified when verifying the session token for precisely this readon. So yes, there would be an unverified token lying around, but that is fine because it might have been created by an attacker, or maybe the user will come back and verify that session later.

But also, I might have misremembered what we thrashed out. @shane-tomlinson, @vbudhram, care to chip in on the above?